### ISSUE TITLE
[Web] if web is hosted in sub path, service worker cache doesn't work.

### ISSUE THREAD

**Issue by JimmyTai** [flutter/flutter#111866]
## Steps to Reproduce

1. Create a sample application project.
2. Build Flutter web and move to a directory.
```shell
flutter build web --release --base-href=/sub/
mkdir -p release/sub && cp -a build/web/* release/sub
```
3. Host Flutter web by http-server. 
```shell
http-server release
```
4. Open Chrome and dev-tools, then load `http:127.0.0.1:8080/sub`
5. At first time loading the web, web will install service_worker and cache files by flutter_service_worker.js CORE and RESOURCES rule. And then refresh the web. You could inspect all network request at Network tab. I expect files like main.dart.js should comes from service_worker rather than disk cache directly.

**Expected results:** 

<img width="1440" alt="截圖 2022-09-19 下午4 39 33" src="https://user-images.githubusercontent.com/14304782/190980580-5a721c16-a9cb-4916-9b1e-53bd2c231c3f.png">

**Actual results:** 

<img width="1440" alt="截圖 2022-09-19 下午4 42 20" src="https://user-images.githubusercontent.com/14304782/190980703-f32f4115-5045-4d42-80d9-33bf90742629.png">

<details>
<summary>Code sample</summary>

```
This issue is related web host bug. It is hard to provide a code sample.
```
</details>

<details>
  <summary>Logs</summary>

```
I think the screenshot is enough.
```

```
Analyzing service_worker_demo...                                
No issues found! (ran in 1.2s)
```

```
[✓] Flutter (Channel stable, 3.3.2, on macOS 12.6 21G115 darwin-arm, locale zh-Hant-TW)
    • Flutter version 3.3.2 on channel stable at /Users/jimmy/fvm/versions/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision e3c29ec00c (5 days ago), 2022-09-14 08:46:55 -0500
    • Engine revision a4ff2c53d8
    • Dart version 2.18.1
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/jimmy/Library/Android/sdk
    • Platform android-32, build-tools 32.1.0-rc1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13C100
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] VS Code (version 1.71.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.48.0

[✓] Connected device (2 available)
    • JimmyTai-iPhone (mobile) • 053fa2bd7d1c65719ac672c401f6c823ccb4b614 • ios            • iOS 15.3.1 19D52
    • Chrome (web)             • chrome                                   • web-javascript • Google Chrome 105.0.5195.125

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by exaby73** [comment]
Hello @JimmyTai. I tried this with python as an http server and could not reproduce it. The command I ran was `python3 -m http.server --directory release`.

### Flutter Doctor

<details><summary><code>flutter doctor -v</code> (Stable)</summary>

```
[✓] Flutter (Channel stable, 3.3.2, on macOS 12.5.1 21G83 darwin-arm, locale en-US)
    • Flutter version 3.3.2 on channel stable at /Users/nabeelparkar/fvm/versions/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision e3c29ec00c (5 days ago), 2022-09-14 08:46:55 -0500
    • Engine revision a4ff2c53d8
    • Dart version 2.18.1
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nabeelparkar/Library/Android/sdk/
    • Platform android-33, build-tools 33.0.0
    • ANDROID_SDK_ROOT = /Users/nabeelparkar/Library/Android/sdk/
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A309
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /Applications/Brave Browser.app/Contents/MacOS/Brave Browser

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Ultimate Edition (version 2022.1.3)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 69.0.4
    • Dart plugin version 221.5921.27

[✓] VS Code (version 1.69.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.44.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 12.5.1 21G83 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Brave Browser 105.1.43.93

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by JimmyTai** [comment]
Let me try again with python http.server.

**Comment by JimmyTai** [comment]
I tried it again.

At first time, `main.dart.js` was fetched from http directly.

<img width="1470" alt="截圖 2022-09-20 下午4 26 21" src="https://user-images.githubusercontent.com/14304782/191208003-3440d53f-3776-4bef-8a5b-9499fbeb5c7a.png">

After refresh the web, `main.dart.js` was fetched via disk cache. If service worker worked, it should be fetched via service worker and return service worker cache then fetch again from http.

<img width="1470" alt="截圖 2022-09-20 下午4 26 35" src="https://user-images.githubusercontent.com/14304782/191208669-a3972ec1-fc8c-41a8-9e41-cfc73efb04d6.png">

**Comment by exaby73** [comment]
Could you check in the `Application` tab in the dev tools if the service worker is registered?

**Comment by JimmyTai** [comment]
@exaby73 service worker is registered, but the cache strategy work incorrectly. Due to the subpath doesn’t include in `origin` like what I mentioned https://github.com/flutter/flutter/issues/111866#issuecomment-1250742648.

In `flutter_service_worker.js` “fetch” listener, it substring `event.request.url` by `self.location.origin` but the key in `RESOURCES` is like this.
```dart
const RESOURCES = {
  "version.json": "2a9e876631b714d5852610883aa93145",
  "index.html": "3631aba9d39d6da69f717c2fd00bb46f",
  "/": "3631aba9d39d6da69f717c2fd00bb46f",
  "main.dart.js": "de7c78bb99078a6edeaece18e1b02cf9",
  "flutter.js": "f85e6fb278b0fd20c349186fb46ae36d",
  "favicon.png": "5dcef449791fa27946b3d35ad8803796",
  "icons/Icon-192.png": "ac9a721a12bbc803b44f645561ecb1e1",
  "icons/Icon-maskable-192.png": "c457ef57daa1d16f64b27b786ec2ea3c",
  "icons/Icon-maskable-512.png": "301a7604d45b3e739efc881eb04896ea",
  "icons/Icon-512.png": "96e752610906ba2a93c65f8abe1645f1",
  "manifest.json": "fdb74b9d1f2e76d61e8ac98ff245d08d",
  "assets/AssetManifest.json": "2efbb41d7877d10aac9d091f58ccd7b9",
  "assets/NOTICES": "fb9e9695eff13a72ecc9316559bef51f",
  "assets/FontManifest.json": "dc3d03800ccca4601324923c0b1d6d57",
  "assets/packages/cupertino_icons/assets/CupertinoIcons.ttf": "6d342eb68f170c97609e9da345464e5e",
  "assets/shaders/ink_sparkle.frag": "01defb711152914bc7fc69040c7f1d01",
  "assets/fonts/MaterialIcons-Regular.otf": "95db9098c58fd6db106f1116bae85a0b",
  "canvaskit/canvaskit.js": "2bc454a691c631b07a9307ac4ca47797",
  "canvaskit/profiling/canvaskit.js": "38164e5a72bdad0faa4ce740c9b8e564",
  "canvaskit/profiling/canvaskit.wasm": "95a45378b69e77af5ed2bc72b2209b94",
  "canvaskit/canvaskit.wasm": "bf50631470eb967688cca13ee181af62"
};
```
### For example
Assume the web is host on `http://localhost:8000/sub/#/` and the main.dart.js will locate at `http://localhost:8000/sub/main.dart.js`

In "fetch" listener, 
`event.request.url` will be `http://localhost:8000/sub/main.dart.js`
`self.location.origin` will be `http://localhost:8000`

After this code in `flutter_service_worker.js`
```dart
var key = event.request.url.substring(origin.length + 1);
```
`key` will be `sub/main.dart.js`
It doesn't match any key in `RESOURCES`, then service worker won't handle `main.dart.js` anymore.

**Comment by exaby73** [comment]
Understood. Thank you for the detailed information. Labelling this issue for further insights from the team.

**Comment by JimmyTai** [comment]
@ditman I am preparing a PR for this issue. 

Could I pass the `baseHref` argument to `WebServiceWorker` class in `compile.dart` line 50 and then pass to `generateServiceWorker()` function in `web.dart` line 506?

Then appends `baseHref` after `self.location.origin` in flutter_service_worker_js.dart line 78 and line 117.

**Comment by ditman** [comment]
@JimmyTai interesting. I thought the service worker could use window.location to determine what's the current base URL, and use that to compose the keys? That way it wouldn't need the "base" at compile time? (maybe `window.location.pathname`?)

**Comment by davidlanderos-digital** [comment]
Hello, I  didn't know about this issue and accidentally [duplicated it](https://github.com/flutter/flutter/issues/115250).

Take a look at my [PR](https://github.com/flutter/flutter/pull/115251), I'm ensuring the RESOURCES array is generated with the full path, I believe this approach is easier to test, (see unit test added)

imho, changing the runtime behavior of the service worker would require a more sophisticated integration test

**Comment by davidlanderos-digital** [comment]
mmh... ok, I'm seeing that the [docs encourage you to edit the HTML](https://docs.flutter.dev/development/ui/navigation/url-strategies) rather than setting it at build time.

**Comment by davidlanderos-digital** [comment]
After reading your discussion I felt guilty for requiring the developer to define base-href at build time.
@JimmyTai are you still working on this?

**Comment by ditman** [comment]
@davidlanderos-digital have we merged the solution to this? I lost track of the PR!

**Comment by kevmoo** [comment]
https://github.com/davidlanderos-digital/flutter/commit/0530063d41ba8d7ea668290cff4cca1dc7ff98b7 right?

**Comment by davidlanderos-digital** [comment]
This is not fixed yet.

Just to clarify, there were 2 different issues:
- service worker not being  able to cache things right after the first launch (Fixed and merged)
- service worker not being able to cache **at all** when base-href is non root. (This one)

Now that the first issue was merged, I believe there is nothing else blocking me from sending a PR for solving this one. and Yes @kevmoo that would be the solution + its e2d tests

**Comment by kevmoo** [comment]
Thanks for your help here, @davidlanderos-digital !

**Comment by ditman** [comment]
Yes! Do ping me in the PR, please! Since this has to do with the `base` path, @mdebbar might be interested too!

**Comment by fabiocody** [comment]
Hello everyone
Any news on this?

**Comment by tuotau-lessonapp** [comment]
> Hello everyone Any news on this?

Seconding this! Almost the first anniversary for this issue :D

**Comment by PiotrFLEURY** [comment]
Hi everyone.

I spent some time to try to understand the full story of this issue.
For those who wants the summary, this is what I understood :

Flutter Web service worker does not cache resources if the Web PWA is deployed under a subdirectory using `base-href`

@davidlanderos-digital first opened [this PR](https://github.com/flutter/flutter/pull/115251) that was closed to try a [better approach](https://github.com/davidlanderos-digital/flutter/commit/0530063d41ba8d7ea668290cff4cca1dc7ff98b7)

Then, [a new PR](https://github.com/flutter/flutter/pull/121348) has been opened to try to fix it. This one has also been close as stale.

As far as I understand, no one is working on this fix anymore. 

Am I right ?

I am a bit confused about the potential fix for this cache problem.
If we do agree for a clear fix I can try to handle this issue and propose an new PR.

Can someone tell me what should be the better approach to fix this ?

- Include `base-href` in cached resource keys ?
- Change key name on the fly in service worker cache resource lookup code ?
- Another ?

**Comment by dg76** [comment]
I would also be interested in an official solution. I had published my own solution (which works fine for me), with code  here: https://github.com/flutter/flutter/issues/129905

However, it was closed as duplicate. But one the other hand there still does not seem to be an official solution.

**Comment by andynewman10** [comment]
@dg76 @JimmyTai @david-landeros Your patches do not work on my codebase. I too am in a situation where my Flutter application is in a subfolder.

Did it really work for you?

For some reason, I never get the `[from service worker]` (see the first screenshot at the top of this issue, Dev Tools right pane) origin in Chrome Dev Tools when contents are retrieved, when refreshing the page (second landing).

Flutter's flutter_service_worker.js code hasn't changed since your various PRs. I have no clue as to what's going on. I am surprised this problem doesn't get more attention as it seriously messes browser caching, and therefore prevents users from getting the latest version of the web application.

**Comment by david-landeros** [comment]
Hello @andynewman10

Did you try https://github.com/flutter/flutter/pull/115251 or https://github.com/flutter/flutter/pull/121348?

The first modifies flutter source code and If I remember correctly for it to take effect you need to delete: **/flutter/bin/cache/flutter_web_sdk.stamp** before re-compiling

The second approach modifies the service_worker directly, if you applied this one I cannot see how it may fail, do you mind sharing your resulting service_worker.js in the build folder?

**Comment by andynewman10** [comment]
@david-landeros it works, I have further customized the code and it was a mistake in my code.
